rasa_sdk/knowledge_base/actions.py

```diff
@@ -0,0 +1,290 @@
+# -*- coding: utf-8 -*-
+from rasa_sdk import Action
+from rasa_sdk.events import SlotSet
+from rasa_sdk.knowledge_base.storage import (
+    SCHEMA_KEYS_ATTRIBUTES,
+    SCHEMA_KEYS_REPRESENTATION,
+    SCHEMA_KEYS_KEY,
+)
+
+
+SLOT_MENTION = "mention"
+SLOT_ENTITY_TYPE = "entity_type"
+SLOT_ATTRIBUTE = "attribute"
+SLOT_LISTED_ITEMS = "knowledge_base_listed_items"
+SLOT_LAST_ENTITY = "knowledge_base_last_entity"
+SLOT_LAST_ENTITY_TYPE = "knowledge_base_last_entity_type"
+
+
+class ActionKnowledgeBase(Action):
+    """
+    Abstract knowledge base action that can be inherited to create custom actions
+    that are able to interact with the knowledge base.
+    """
+
+    def __init__(self, knowledge_base):
+        self.knowledge_base = knowledge_base
+
+    def _get_entity(self, tracker):
+        """
+        Get the name of the entity the user referred to. Either the NER detected the
+        entity and stored its name in the corresponding slot or the user referred to
+        the entity by any kind of mention, such as "first one" or "it".
+
+        Args:
+            tracker: the tracker
+
+        Returns: the name of the actual entity (value of key attribute in the knowledge base)
+        """
+        mention = tracker.get_slot(SLOT_MENTION)
+        entity_type = tracker.get_slot(SLOT_ENTITY_TYPE)
+
+        if mention:
+            return self._resolve_mention(tracker)
+
+        # check whether the user referred to the entity by its name
+        entity_name = tracker.get_slot(entity_type)
+        if entity_name:
+            return entity_name
+
+        # if no explicit mention was found, we assume the user just refers to the last
+        # entity mentioned in the conversation
+        return tracker.get_slot(SLOT_LAST_ENTITY)
+
+    def _resolve_mention(self, tracker):
+        """
+        Resolves a mention of an entity, such as first, to the actual entity.
+        If multiple entities are listed during the conversation, the entities
+        are stored in the slot 'knowledge_base_entities' as a list. We resolve the
+        mention, such as first, to the list index and retrieve the actual entity.
+        If the mention is not an ordinal mention, but some other reference, we just
+        assume the user is referring to the last mentioned entity in the conversation.
```

Comment:
following on from our discussion, I feel we are potentially causing a lot of confusion by referring to the objects in the knowledge base as entities. Maybe we should call them something like 'objects' or 'items' or 'nodes' or 'mentions'

Reply:
changed `entities` to `objects`